doc/source/templates/05_dreambooth_finetuning/dreambooth/train.py

```diff
@@ -77,70 +140,78 @@ def load_models(config, cuda):
     )
     # We are not training VAE part of the model.
     vae.requires_grad_(False)
-    vae.to(cuda[1])
 
     # Convert unet to bf16 to save GRAM.
     unet = UNet2DConditionModel.from_pretrained(
         config["model_dir"],
         subfolder="unet",
         torch_dtype=dtype,
     )
+
     if is_xformers_available():
         unet.enable_xformers_memory_efficient_attention()
-    # UNET is the largest component, occupying first GPU by itself.
-    unet.to(cuda[0])
-    unet.train()
-
-    torch.cuda.empty_cache()
 
-    return text_encoder, noise_scheduler, vae, unet
+    if not config["use_lora"]:
+        unet_parameters = unet.parameters()
+        text_parameters = text_encoder.parameters()
+    else:
+        text_encoder.requires_grad_(False)
+        unet.requires_grad_(False)
+        unet_parameters, text_parameters = add_lora_layers(unet, text_encoder)
 
+    torch.cuda.empty_cache()
 
-def get_cuda_devices():
-    devices = [f"cuda:{i}" for i in range(torch.cuda.device_count())]
-    local_rank = train.get_context().get_local_rank()
-    assert len(devices) >= 2, "Require at least 2 GPU devices to work."
-    return devices[(local_rank * 2) : ((local_rank * 2) + 2)]
+    return text_encoder, noise_scheduler, vae, unet, unet_parameters, text_parameters
 
 
 def train_fn(config):
-    cuda = get_cuda_devices()
 
     # Load pre-trained models.
-    text_encoder, noise_scheduler, vae, unet = load_models(config, cuda)
+    (
+        text_encoder,
+        noise_scheduler,
+        vae,
+        unet,
+        unet_parameters,
+        text_parameters,
+    ) = load_models(config)
 
-    # Wrap in DDP
-    text_encoder = DistributedDataParallel(
-        text_encoder, device_ids=[cuda[1]], output_device=cuda[1]
-    )
-    unet = DistributedDataParallel(unet, device_ids=[cuda[0]], output_device=cuda[0])
+    text_encoder.train()
+    unet.train()
+
+    text_encoder = train.torch.prepare_model(text_encoder)
+    unet = train.torch.prepare_model(unet)
+    vae = vae.to(train.torch.get_device())
```

Comment:
As a user, I need clarification. So what does `train.torch.get_device()` return in this case? is it cuda:1? Why did we not have to do it for `unet`? Also for data why do we do `to(cuda)`?

Reply:
Can we leave comments for this explanation?